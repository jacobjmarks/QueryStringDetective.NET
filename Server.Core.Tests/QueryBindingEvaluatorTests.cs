namespace Server.Core.Tests;

public class QueryBindingEvaluatorTests
{
    public static readonly TheoryData<string> TestCases = new()
    {
        "c",
        "foo",
        "a&q=b&q=c",
        "null",
        "0",
        "1",
        "-1",
        "1&q=2&q=3",
        "1&q=2&q=.3",
        "-1&q=0&q=1",
        $"{sbyte.MinValue}",
        $"{sbyte.MaxValue}",
        $"{byte.MaxValue}",
        $"{short.MinValue}",
        $"{short.MaxValue}",
        $"{ushort.MaxValue}",
        $"{int.MinValue}",
        $"{int.MaxValue}",
        $"{long.MinValue}",
        $"{long.MaxValue}",
        $"{ulong.MaxValue}",
        Uri.EscapeDataString($"{double.MinValue}"),
        Uri.EscapeDataString($"{double.MaxValue}"),
        $"{decimal.MinValue}",
        $"{decimal.MaxValue}",
        Uri.EscapeDataString($"{float.MinValue}"),
        Uri.EscapeDataString($"{float.MaxValue}"),
    };

    [Theory]
    [MemberData(nameof(TestCases))]
    public async Task AssertBinding(string queryString)
    {
        using var bindingEvaluator = new QueryBindingEvaluator();
        var results = await bindingEvaluator.EvaluateAsync("?q=" + queryString);

        var content = "Input: " + queryString
            + "\n\nBindings:\n" + string.Join("\n", results.Where(r => !r.IsErroneous)
                    .Select(r => $"{r.Type,-10} {r.Result}"))
            + "\n\nErroneous:\n" + string.Join("\n", results.Where(r => r.IsErroneous)
                    .Select(r => $"{r.Type,-10} {r.Error!.Message}"));

        content.Should().MatchSnapshot(SnapshotNameExtension.Create(ToValidFileName(queryString)));
    }

    private static string ToValidFileName(string value)
    {
        foreach (var invalidChar in Path.GetInvalidFileNameChars())
            value = value.Replace(invalidChar, '_');

        return value;
    }
}