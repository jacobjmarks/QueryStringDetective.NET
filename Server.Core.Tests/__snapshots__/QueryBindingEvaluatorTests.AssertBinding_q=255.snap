Input: q=255

Bindings:

string
  MinimalApis
    r = "255"
  Controllers
    r = "255"

string?
  MinimalApis
    r = "255"
  Controllers
    r = "255"

string[]
  MinimalApis
    r = ["255"]
  Controllers
    r = ["255"]

char
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "char q" from "255".)
  Controllers
    e = 400 Bad Request (The value '255' is not valid.)

char?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<char> q" from "255".)
  Controllers
    e = 400 Bad Request (The value '255' is not valid.)

char[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "char[] q" from "255".)
  Controllers
    e = 400 Bad Request (The value '255' is not valid.)

int
  MinimalApis
    r = 255
  Controllers
    r = 255

int?
  MinimalApis
    r = 255
  Controllers
    r = 255

int[]
  MinimalApis
    r = [255]
  Controllers
    r = [255]

double
  MinimalApis
    r = 255
  Controllers
    r = 255

double?
  MinimalApis
    r = 255
  Controllers
    r = 255

double[]
  MinimalApis
    r = [255]
  Controllers
    r = [255]

float
  MinimalApis
    r = 255
  Controllers
    r = 255

float?
  MinimalApis
    r = 255
  Controllers
    r = 255

float[]
  MinimalApis
    r = [255]
  Controllers
    r = [255]

decimal
  MinimalApis
    r = 255
  Controllers
    r = 255

decimal?
  MinimalApis
    r = 255
  Controllers
    r = 255

decimal[]
  MinimalApis
    r = [255]
  Controllers
    r = [255]

bool
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "bool q" from "255".)
  Controllers
    e = 400 Bad Request (The value '255' is not valid.)

bool?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<bool> q" from "255".)
  Controllers
    e = 400 Bad Request (The value '255' is not valid.)

bool[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "bool[] q" from "255".)
  Controllers
    e = 400 Bad Request (The value '255' is not valid.)

uint
  MinimalApis
    r = 255
  Controllers
    r = 255

uint?
  MinimalApis
    r = 255
  Controllers
    r = 255

uint[]
  MinimalApis
    r = [255]
  Controllers
    r = [255]

long
  MinimalApis
    r = 255
  Controllers
    r = 255

long?
  MinimalApis
    r = 255
  Controllers
    r = 255

long[]
  MinimalApis
    r = [255]
  Controllers
    r = [255]

ulong
  MinimalApis
    r = 255
  Controllers
    r = 255

ulong?
  MinimalApis
    r = 255
  Controllers
    r = 255

ulong[]
  MinimalApis
    r = [255]
  Controllers
    r = [255]

sbyte
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "sbyte q" from "255".)
  Controllers
    e = 400 Bad Request (The value '255' is not valid.)

sbyte?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<sbyte> q" from "255".)
  Controllers
    e = 400 Bad Request (The value '255' is not valid.)

sbyte[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "sbyte[] q" from "255".)
  Controllers
    e = 400 Bad Request (The value '255' is not valid.)

short
  MinimalApis
    r = 255
  Controllers
    r = 255

short?
  MinimalApis
    r = 255
  Controllers
    r = 255

short[]
  MinimalApis
    r = [255]
  Controllers
    r = [255]

ushort
  MinimalApis
    r = 255
  Controllers
    r = 255

ushort?
  MinimalApis
    r = 255
  Controllers
    r = 255

ushort[]
  MinimalApis
    r = [255]
  Controllers
    r = [255]

byte
  MinimalApis
    r = 255
  Controllers
    r = 255

byte?
  MinimalApis
    r = 255
  Controllers
    r = 255

byte[]
  MinimalApis
    r = "/w=="
  Controllers
    e = 400 Bad Request (The value '255' is not valid.)
